### PR TITLE
New resources: `azurerm_redis_cache_policy` and `azurerm_redis_cache_policy_assignment`

### LINKED ISSUE #24692: Support for Redis User
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

On the `(PREVIEW) Data Access Configuration` tab of Azure Cache for Redis, there is the ability to add Redis users and assign them different access policies, e.g. Data Owner. Can this feature be added to Terraform?

### New or Affected Resource(s)/Data Source(s)

azurerm_redis_cache

### Potential Terraform Configuration

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by favoretti**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## Description

Adds 2 new resources:
`azurerm_redis_cache_access_policy`
`azurerm_redis_cache_access_policy_assignment`

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

image

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* New resource: `azurerm_redis_cache_access_policy` 
* New resource: `azurerm_redis_cache_access_policy_assignment` 
* `azurerm_redis_cache`: Up Create/Update/Delete default timeouts to 180 minutes

This is a (please select all that apply):

- [x] New Feature (ie adding a service, resource, or data source)

## Related Issue(s)
Fixes #24692

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/redis/redis_cache_access_policy_assignment_resource.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+	return map[string]*pluginsdk.Schema{
+		"name": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+		},
+		"redis_cache_id": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+			ForceNew: true,
+		},
```
Comment (catriona-m) [inline internal/services/redis/redis_cache_access_policy_assignment_resource.go:37]: could we use the redis cache id validation for this?

File: internal/services/redis/redis_cache_access_policy_assignment_resource_test.go
Comment (catriona-m) [inline internal/services/redis/redis_cache_access_policy_assignment_resource_test.go:1]: could we include an update and a requiresImport tests for this?

File: internal/services/redis/redis_cache_access_policy_assignment_resource.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	ObjectID         string `tfschema:"object_id"`
+	ObjectIDAlias    string `tfschema:"object_id_alias"`
+}
+
+func (r RedisCacheAccessPolicyAssignmentResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"name": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+		},
```
Comment (catriona-m) [inline internal/services/redis/redis_cache_access_policy_assignment_resource.go:32]:
```suggestion
		"name": {
			Type:     pluginsdk.TypeString,
			Required: true,
			ValidateFunc: validation.StringIsNotEmpty,
		},
```

File: internal/services/redis/redis_cache_access_policy_resource.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+	return map[string]*pluginsdk.Schema{
+		"name": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+		},
+		"redis_cache_id": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+			ForceNew: true,
+		},
```
Comment (catriona-m) [inline internal/services/redis/redis_cache_access_policy_resource.go:35]: could we add validation to these?

File: internal/services/redis/redis_cache_access_policy_assignment_resource.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	return "azurerm_redis_cache_access_policy_assignment"
+}
+
+func (r RedisCacheAccessPolicyAssignmentResource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return redis.ValidateAccessPolicyAssignmentID
+}
+
+func (r RedisCacheAccessPolicyAssignmentResource) Create() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
```
Comment (catriona-m) [inline internal/services/redis/redis_cache_access_policy_assignment_resource.go:71]: should this be 60 minutes?

File: internal/services/redis/redis_cache_access_policy_assignment_resource.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+			metadata.SetID(id)
+			return nil
+		},
+	}
+}
+
+func (r RedisCacheAccessPolicyAssignmentResource) Update() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
```
Comment (catriona-m) [inline internal/services/redis/redis_cache_access_policy_assignment_resource.go:106]: should this be 60 minutes?

File: internal/services/redis/redis_cache_access_policy_resource.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+	return "azurerm_redis_cache_access_policy"
+}
+
+func (r RedisCacheAccessPolicyResource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return redis.ValidateAccessPolicyID
+}
+
+func (r RedisCacheAccessPolicyResource) Create() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
```
Comment (catriona-m) [inline internal/services/redis/redis_cache_access_policy_resource.go:61]: should this be 60 minutes?

File: internal/services/redis/redis_cache_access_policy_resource.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+			metadata.SetID(id)
+			return nil
+		},
+	}
+}
+
+func (r RedisCacheAccessPolicyResource) Update() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
```
Comment (catriona-m) [inline internal/services/redis/redis_cache_access_policy_resource.go:97]: should this be 60 minutes?

File: internal/services/redis/redis_cache_access_policy_resource.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+	RedisCacheID string `tfschema:"redis_cache_id"`
+	Permissions  string `tfschema:"permissions"`
+}
+
+func (r RedisCacheAccessPolicyResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"name": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+		},
```
Comment (catriona-m) [inline internal/services/redis/redis_cache_access_policy_resource.go:30]: should this be `ForceNew`?

File: website/docs/r/redis_cache_access_policy.html.markdown [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+}
+```
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the Redis Cache Access Policy.
+
+* `redis_cache_id` - (Required) The ID of the Redis Cache. Changing this forces a new Redis to be created.
```
Comment (catriona-m) [inline website/docs/r/redis_cache_access_policy.html.markdown:50]:
```suggestion
* `redis_cache_id` - (Required) The ID of the Redis Cache. Changing this forces a new Redis Cache Access Policy to be created.
```

File: website/docs/r/redis_cache_access_policy.html.markdown [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+* `permissions` - (Required) Permissions that are going to be assigned to this Redis Cache Access Policy.
+
+## Attributes Reference
+
+In addition to the Arguments listed above - the following Attributes are exported:
+
+* `id` - The ID of the Redis.
```
Comment (catriona-m) [inline website/docs/r/redis_cache_access_policy.html.markdown:58]:
```suggestion
* `id` - The ID of the Redis Cache Access Policy.
```

File: website/docs/r/redis_cache_access_policy.html.markdown [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+## Timeouts
+
+The `timeouts` block allows you to specify [timeouts](https://www.terraform.io/language/resources/syntax#operation-timeouts) for certain actions:
+
+* `create` - (Defaults to 60 minutes) Used when creating the Redis Cache Access Policy.
+* `read` - (Defaults to 5 minutes) Used when retrieving the Redis Cache Access Policy.
+* `delete` - (Defaults to 60 minutes) Used when deleting the Redis Cache Access Policy.
```
Comment (catriona-m) [inline website/docs/r/redis_cache_access_policy.html.markdown:66]: could we include the `update` timeout here?

File: website/docs/r/redis_cache_access_policy.html.markdown [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+## Import
+
+Redis can be imported using the `resource id`, e.g.
```
Comment (catriona-m) [inline website/docs/r/redis_cache_access_policy.html.markdown:70]:
```suggestion
Redis Cache Access Policy can be imported using the `resource id`, e.g.
```

File: website/docs/r/redis_cache_access_policy_assignment.html.markdown [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+  object_id          = data.azurerm_client_config.test.object_id
+  object_id_alias    = "ServicePrincipal"
+}
+```
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the Redis Cache Access Policy.
```
Comment (catriona-m) [inline website/docs/r/redis_cache_access_policy_assignment.html.markdown:53]:
```suggestion
* `name` - (Required) The name of the Redis Cache Access Policy Assignment.
```

File: website/docs/r/redis_cache_access_policy_assignment.html.markdown [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+* `redis_cache_id` - (Required) The ID of the Redis Cache. Changing this forces a new Redis to be created.
```
Comment (catriona-m) [inline website/docs/r/redis_cache_access_policy_assignment.html.markdown:55]:
```suggestion
* `redis_cache_id` - (Required) The ID of the Redis Cache. Changing this forces a new Redis Cache Access Policy Assignment to be created.
```

File: internal/services/redis/redis_cache_access_policy_assignment_resource.go [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+			Required: true,
+		},
+		"object_id": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+		},
+		"object_id_alias": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+		},
```
Comment (catriona-m) [inline internal/services/redis/redis_cache_access_policy_assignment_resource.go:49]: could we add validation here for the possible values `ServicePrincipal` and `UserMSI`?

File: website/docs/r/redis_cache_access_policy_assignment.html.markdown [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+
+* `object_id` - (Required) The principal ID to be assigned the Access Policy.
+
+* `object_id_alias` - (Required) The alias of the principal ID. Possible values are `ServicePrincipal` and `UserMSI`.
+
+## Attributes Reference
+
+In addition to the Arguments listed above - the following Attributes are exported:
+
+* `id` - The ID of the Redis.
```
Comment (catriona-m) [inline website/docs/r/redis_cache_access_policy_assignment.html.markdown:67]:
```suggestion
* `id` - The ID of the Redis Cache Access Policy Assignment.
```

File: website/docs/r/redis_cache_access_policy_assignment.html.markdown [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+
+## Timeouts
+
+The `timeouts` block allows you to specify [timeouts](https://www.terraform.io/language/resources/syntax#operation-timeouts) for certain actions:
+
+* `create` - (Defaults to 60 minutes) Used when creating the Redis Cache Access Policy Assignment.
+* `read` - (Defaults to 5 minutes) Used when retrieving the Redis Cache Access Policy Assignment.
+* `delete` - (Defaults to 60 minutes) Used when deleting the Redis Cache Access Policy Assignment.
```
Comment (catriona-m) [inline website/docs/r/redis_cache_access_policy_assignment.html.markdown:75]: could we include the update timeout here?

**Review by catriona-m — CHANGES_REQUESTED** [review]
Hi @favoretti thanks for submitting this! I left a few comments/queries inline, but can take another look through once those are addressed. Thanks!

**Comment by favoretti** [comment]
@catriona-m thank you for the review. Apologies for so many copypaste errors. Coding during meetings is not my strongest treat it seems :( I'll address the tests/timeouts/validations later tonight.

File: internal/services/redis/redis_cache_access_policy_assignment_resource.go (hunk 15 above)
Comment (audunsolemdal) [reply internal/services/redis/redis_cache_access_policy_assignment_resource.go]: @catriona-m I am curious of what the purpose for this validation, I first created the resources with azapi and am now trying to import them to the azurerm provider

`object_id_alias    = "xxx@yyy.com" -> "UserMSI" # forces replacement`

Ref. https://learn.microsoft.com/en-us/azure/templates/microsoft.cache/redis/accesspolicyassignments?pivots=deployment-language-terraform#rediscacheaccesspolicyassignmentproperties-2

Any insights you could share?

Comment (catriona-m) [reply internal/services/redis/redis_cache_access_policy_assignment_resource.go]: Thanks for highlighting this @audunsolemdal, it looks like we'll need to update the validation to account for usernames too.

Comment (audunsolemdal) [reply internal/services/redis/redis_cache_access_policy_assignment_resource.go]: In my case I also set object_id_alias =  <name of my app service>

We are yet to test redis cache policies from these app services though. Not sure how exactly this attribute is intended to be used

Comment (favoretti) [reply internal/services/redis/redis_cache_access_policy_assignment_resource.go]: @audunsolemdal @catriona-m ahh, so it *is* an arbitrary string. I'll add a PR to make it StringNotEmpty validation

**Pull request merged**